[18.0][FIX] account_payment_partner: missing migration scripts

remi-filament opened the pull request:
The fields customer_payment_mode_id and supplier_payment_mode_id on res_partner are company_dependent and therefore need to be migrated in v18.0 in case you come from a previous Odoo version.

This PR converts fields from ir_property.

It should be transparent if properties do not exist.

remi-filament commented:
Hi @carlos-lopez-tecnativa @pedrobaeza would you mind having a look at that one ?
I am not sure if I should use a new version number or set migration scripts on previous version (for instance 18.0.1.0.0) so that it is triggered only when migrating from previous version ?

I chose a new version because in case database was already migrated but old properties were not removed they should still be present in DB and could be recovered by updating this module.

pedrobaeza commented:
On my disputed contribution guideline addition, I propose to put these commits as `[OU-ADD]`.

pedrobaeza commented on account_payment_partner/__manifest__.py:
```diff
@@ -6,7 +6,7 @@
 {
     "name": "Account Payment Partner",
-    "version": "18.0.1.0.1",
+    "version": "18.0.2.0.0",
```
Don't change this, and the migration script should be in `18.0.1.0.0` folder. If you want to trigger this on your migrated DB, execute in a shell or reset the latest version to a previous one.

carlos-lopez-tecnativa commented on account_payment_partner/migrations/18.0.2.0.0/post-migration.py:
```diff
@@ -0,0 +1,14 @@
+
+@openupgrade.migrate()
+def migrate(env, version):
+    openupgrade_180.convert_company_dependent(
```
I see an error and noticed it comes from `openupgradelib`. It does not handle if the table exists.

In my test, I worked with a new database created directly in V18, and later I updated the module with new changes, so ir_property did not exist.

I think this fix must be done in `openupgradelib`, @pedrobaeza WDYT?
https://github.com/OCA/openupgradelib/blob/8a86014577d567883f73e36eea436af67096078d/openupgradelib/openupgrade_180.py#L66

<img width="1084" height="361" alt="image" src="https://github.com/user-attachments/assets/eade224e-6285-4087-9422-9b71ea0f45f4" />

pedrobaeza replied:
This can't happen, as `ir_property` must exist when coming from 17.0. If you have already purge obsolete data (through `database_cleanup`), then you can't recover this information. That's why I recommend no to clean things after certain time (or even only before starting the next migration cycle).

And another reason to not bump the manifest version here.

carlos-lopez-tecnativa replied:
> This can't happen, as `ir_property` must exist when coming from 17.0. If you have already purge obsolete data (through `database_cleanup`), then you can't recover this information. That's why I recommend no to clean things after certain time (or even only before starting the next migration cycle).
> 
> And another reason to not bump the manifest version here.

Based on this, this script should only be executed when migrating a database from V17 to V18, not as a regular migration script. So @remi-filament please do not bump the manifest version. and put  the migration in the `18.0.1.0.0` folder. Otherwise, in the current databases created from V18 (not migrated), updating this module will cause the error.

pedrobaeza replied:
Already said in https://github.com/OCA/bank-payment/pull/1492#discussion_r2276905376

pedrobaeza commented on account_payment_partner/README.rst:
```diff
@@ -30,12 +30,12 @@ Account Payment Partner
-- the *# of digits for customer bank account* on Payment Mode.
-- the *Bank account from journals* on Payment Mode.
-- the *Payment mode* on Invoices Analysis.
+-  the *Supplier Payment Mode* and *Customer Payment Mode* on Partners,
```
You should revert these changes.

remi-filament replied:
That was done automagically by pre-commit, I can revert those but pre-commit will probably fail...

remi-filament committed:
[OU-ADD] account_payment_partner: add migration script for v18.0

remi-filament commented:
Strange, pre-commit is still OK without README changes.
I updated version to 18.0.1.0.0, removed change on __manifest__.py and renamed commit as per suggestion ([OU-ADD])

Thanks for your review @pedrobaeza @carlos-lopez-tecnativa

matteoopenf commented:
hi,

during migration on odoo.sh I obtain that when this migration start

ERROR mydb-18-0-mig-branch-23619404 odoo.sql_db: bad query: b'\n        UPDATE res_partner SET supplier_payment_mode_id=ir_property_by_company.value\n        FROM (\n            SELECT\n            SPLIT_PART(res_id, \',\', 2)::integer res_id,\n            JSON_OBJECT_AGG(company_id, SPLIT_PART(value_reference, \',\', 2)::integer) AS "value"\n            FROM ir_property\n            WHERE\n            fields_id=18550 AND res_id IS NOT NULL\n            AND company_id IS NOT NULL\n            GROUP BY res_id\n        ) ir_property_by_company\n        WHERE res_partner.id=ir_property_by_company.res_id\n        '
ERROR: relation "ir_property" does not exist
LINE 7:             FROM ir_property
                         ^
2025-09-11 07:08:02,540 38 ERROR mydb-18-0-mig-branch-23619404 OpenUpgrade: Error after 0:00:00.017899 running 
        UPDATE res_partner SET supplier_payment_mode_id=ir_property_by_company.value
        FROM (
            SELECT
            SPLIT_PART(res_id, ',', 2)::integer res_id,
            JSON_OBJECT_AGG(company_id, SPLIT_PART(value_reference, ',', 2)::integer) AS "value"
            FROM ir_property
            WHERE
            fields_id=18550 AND res_id IS NOT NULL
            AND company_id IS NOT NULL
            GROUP BY res_id
        ) ir_property_by_company
        WHERE res_partner.id=ir_property_by_company.res_id
2025-09-11 07:08:02,542 38 ERROR mydb--18-0-mig-branch-23619404 OpenUpgrade: account_payment_partner: error in migration script /home/odoo/src/user/b_OCA/bank-payment/account_payment_partner/migrations/18.0.1.0.0/post-migration.py: relation "ir_property" does not exist
LINE 7:             FROM ir_property
                         ^
2025-09-11 07:08:02,543 38 ERROR mydb-aropenforce-18-0-mig-branch-23619404 OpenUpgrade: relation "ir_property" does not exist
LINE 7:             FROM ir_property
                         ^
 
Traceback (most recent call last):
  File "/home/odoo/.local/lib/python3.12/site-packages/openupgradelib/openupgrade.py", line 2400, in wrapped_function
    func(
  File "/home/odoo/src/user/b_OCA/bank-payment/account_payment_partner/migrations/18.0.1.0.0/post-migration.py", line 9, in migrate
    openupgrade_180.convert_company_dependent(
  File "/home/odoo/.local/lib/python3.12/site-packages/openupgradelib/openupgrade_180.py", line 46, in convert_company_dependent
    logged_query(
  File "/home/odoo/.local/lib/python3.12/site-packages/openupgradelib/openupgrade.py", line 1750, in logged_query
    cr.execute(query, args)
  File "/home/odoo/src/odoo/odoo/sql_db.py", line 357, in execute
    res = self._obj.execute(query, params)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
psycopg2.errors.UndefinedTable: relation "ir_property" does not exist
LINE 7:             FROM ir_property
                         ^
2025-09-11 07:08:02,548 38 WARNINGmydb--18-0-mig-branch-23619404 odoo.modules.loading: Transient module states were reset
2025-09-11 07:08:02,550 38 ERROR mydb--18-0-mig-branch-23619404 odoo.modules.registry: Failed to load registry
2025-09-11 07:08:02,552 38 CRITICAL mydb--18-0-mig-branch-23619404 odoo.service.server: Failed to initialize database `mydb--18-0-mig-branch-23619404`. 
Traceback (most recent call last):
  File "/home/odoo/src/odoo/odoo/service/server.py", line 1361, in preload_registries
    registry = Registry.new(dbname, update_module=update_module)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/decorator.py", line 232, in fun
    return caller(func, *(extras + args), **kw)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/odoo/src/odoo/odoo/tools/func.py", line 97, in locked
    return func(inst, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/odoo/src/odoo/odoo/modules/registry.py", line 129, in new
    odoo.modules.load_modules(registry, force_demo, status, update_module)
  File "/home/odoo/src/odoo/odoo/modules/loading.py", line 485, in load_modules
    processed_modules += load_marked_modules(env, graph,
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/odoo/src/odoo/odoo/modules/loading.py", line 365, in load_marked_modules
    loaded, processed = load_module_graph(
                        ^^^^^^^^^^^^^^^^^^
  File "/home/odoo/src/odoo/odoo/modules/loading.py", line 233, in load_module_graph
    migrations.migrate_module(package, 'post')
  File "/home/odoo/src/odoo/odoo/modules/migration.py", line 222, in migrate_module
    exec_script(self.cr, installed_version, pyfile, pkg.name, stage, stageformat[stage] % version)
  File "/home/odoo/src/odoo/odoo/modules/migration.py", line 259, in exec_script
    mod.migrate(cr, installed_version)
  File "/home/odoo/.local/lib/python3.12/site-packages/openupgradelib/openupgrade.py", line 2400, in wrapped_function
    func(
  File "/home/odoo/src/user/b_OCA/bank-payment/account_payment_partner/migrations/18.0.1.0.0/post-migration.py", line 9, in migrate
    openupgrade_180.convert_company_dependent(
  File "/home/odoo/.local/lib/python3.12/site-packages/openupgradelib/openupgrade_180.py", line 46, in convert_company_dependent
    logged_query(
  File "/home/odoo/.local/lib/python3.12/site-packages/openupgradelib/openupgrade.py", line 1750, in logged_query
    cr.execute(query, args)
  File "/home/odoo/src/odoo/odoo/sql_db.py", line 357, in execute
    res = self._obj.execute(query, params)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
psycopg2.errors.UndefinedTable: relation "ir_property" does not exist
LINE 7:             FROM ir_property
                         ^
2025-09-11 07:08:02,554 38 I

pedrobaeza commented:
It seems Odoo migration engine is dropping `ir_property` table, so how it's supposed you are going to rescue the values of these properties? This can be considered a fault of Odoo, not of this.

matteoopenf commented:
> It seems Odoo migration engine is dropping `ir_property` table, so how it's supposed you are going to rescue the values of these properties? This can be considered a fault of Odoo, not of this.

we try to ask to support to keep the table

jdidderen commented:
> It seems Odoo migration engine is dropping `ir_property` table, so how it's supposed you are going to rescue the values of these properties? This can be considered a fault of Odoo, not of this.

@pedrobaeza @matteoopenf The table is not dropped but is renamed to _ir_property and the data is migrated by Odoo migration scripts. 

This is the result for a database from 13.0 to 18.0
<img width="312" height="460" alt="image" src="https://github.com/user-attachments/assets/65da94a8-0b30-4864-afba-f93e10e0fb8b" />


I have the same error on my side but not really a blocking issue. 

@pedrobaeza Would it be fine to adapt the post-migration to check if the ir_property table exists ?

pedrobaeza commented:
I don't think so, as silently avoid the crash won't fill the needed information, resulting in a non correct migration, and adding logic for how enterprise handles this will complicate the migration scripts and very prone to errors, as the result is not verifiable.

desdelinux commented:
How have you solved this case @pedrobaeza @matteoopenf ?:

` odoo.sql_db: bad query: b'\n        UPDATE res_partner SET supplier_payment_mode_id=ir_property_by_company.value\n        FROM (\n            SELECT\n            SPLIT_PART(res_id, \',\', 2)::integer res_id,\n            JSON_OBJECT_AGG(company_id, SPLIT_PART(value_reference, \',\', 2)::integer) AS "value"\n            FROM ir_property\n            WHERE\n            fields_id=9922 AND res_id IS NOT NULL\n            AND company_id IS NOT NULL\n            GROUP BY res_id\n        ) ir_property_by_company\n        WHERE res_partner.id=ir_property_by_company.res_id\n        '
ERROR: relation "ir_property" does not exist
LINE 7:             FROM ir_property
                         ^
2025-10-01 16:19:02,579 648 ERROR upgrade-desdelinux-24191176 OpenUpgrade: Error after 0:00:00.000265 running 
        UPDATE res_partner SET supplier_payment_mode_id=ir_property_by_company.value
        FROM (
            SELECT
            SPLIT_PART(res_id, ',', 2)::integer res_id,
            JSON_OBJECT_AGG(company_id, SPLIT_PART(value_reference, ',', 2)::integer) AS "value"
            FROM ir_property
            WHERE
            fields_id=9922 AND res_id IS NOT NULL
            AND company_id IS NOT NULL
            GROUP BY res_id
        ) ir_property_by_company
        WHERE res_partner.id=ir_property_by_company.res_id
2025-10-01 16:19:02,579 648 ERROR upgrade-desdelinux-24191176 OpenUpgrade: account_payment_partner: error in migration script /home/odoo/src/user/OCA/bank-payment/account_payment_partner/migrations/18.0.1.0.0/post-migration.py: relation "ir_property" does not exist
LINE 7:             FROM ir_property
                         ^
2025-10-01 16:19:02,579 648 ERROR upgrade-desdelinux-24191176 OpenUpgrade: relation "ir_property" does not exist
LINE 7:             FROM ir_property
                         ^
 
Traceback (most recent call last):
  File "/home/odoo/.local/lib/python3.12/site-packages/openupgradelib/openupgrade.py", line 2400, in wrapped_function
    func(
  File "/home/odoo/src/user/OCA/bank-payment/account_payment_partner/migrations/18.0.1.0.0/post-migration.py", line 9, in migrate
    openupgrade_180.convert_company_dependent(
  File "/home/odoo/.local/lib/python3.12/site-packages/openupgradelib/openupgrade_180.py", line 46, in convert_company_dependent
    logged_query(
  File "/home/odoo/.local/lib/python3.12/site-packages/openupgradelib/openupgrade.py", line 1750, in logged_query
    cr.execute(query, args)
  File "/home/odoo/src/odoo/odoo/sql_db.py", line 357, in execute
    res = self._obj.execute(query, params)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
psycopg2.errors.UndefinedTable: relation "ir_property" does not exist
LINE 7:             FROM ir_property
                         ^`

pedrobaeza commented:
That's the price of going with enteprise, @desdelinux: you need to adapt the OCA migration scripts for working on your already migrated DB to final version provided by enterprise service. And I would say more: this adaptation may not be the same if your source version is one or another. For example, one adapted script may work if coming from version 15, but not if coming from version 16.

matteoopenf commented:
We solved to Ask to support to keep the table ir_property, other is to make a copy of the table and use it during the migration
